kroxylicious-authorizer-providers/kroxylicious-authorizer-acl/src/main/java/io/kroxylicious/authorizer/provider/acl/AclAuthorizer.java

```diff
@@ -0,0 +1,466 @@
+                                                        Subject subject,
+                                                        Action action,
+                                                        Decision whenFound,
+                                                        @Nullable Decision whenNotFound) {
+        assert (whenFound != whenNotFound);
+        for (var p : subject.principals()) {
+
+            Decision foundDecision;
+            var grant = allowPerPrincipal.lookup(p.getClass(), TypeNameMap.Predicate.TYPE_EQUAL_NAME_EQUAL, p.name());
+            if (grant != null) {
+                foundDecision = getDecision(action, grant, whenFound);
+                if (foundDecision != null) {
+                    return foundDecision;
+                }
+            }
+            grant = allowPerPrincipal.lookup(p.getClass(), TypeNameMap.Predicate.TYPE_EQUAL_NAME_ANY, null);
+            if (grant != null) {
+                foundDecision = getDecision(action, grant, whenFound);
+                if (foundDecision != null) {
+                    return foundDecision;
+                }
+            }
+            grant = allowPerPrincipal.lookup(p.getClass(), TypeNameMap.Predicate.TYPE_EQUAL_NAME_STARTS_WITH, p.name());
+            if (grant != null) {
+                foundDecision = getDecision(action, grant, whenFound);
+                if (foundDecision != null) {
+                    return foundDecision;
+                }
+            }
+        }
+        return whenNotFound;
+    }
+
+    @Nullable
+    private static Decision getDecision(Action action,
+                                        PrincipalGrants grants,
+                                        Decision whenFound) {
+        Set<? extends ResourceType<?>> operations;
+        var typeNameMap = grants.nameMatches();
+        ResourceType<?> resourceType = action.operation();
+        if (typeNameMap != null) {
+            operations = typeNameMap.lookup(action.resourceTypeClass(),
+                    TypeNameMap.Predicate.TYPE_EQUAL_NAME_EQUAL,
+                    action.resourceName());
+            if (isFound(operations, resourceType)) {
+                return whenFound;
+            }
+            operations = typeNameMap.lookup(action.resourceTypeClass(),
+                    TypeNameMap.Predicate.TYPE_EQUAL_NAME_ANY,
+                    null);
+            if (isFound(operations, resourceType)) {
+                return whenFound;
+            }
+            operations = typeNameMap.lookup(action.resourceTypeClass(),
+                    TypeNameMap.Predicate.TYPE_EQUAL_NAME_STARTS_WITH,
+                    action.resourceName());
+            if (isFound(operations, resourceType)) {
+                return whenFound;
+            }
+        }
+        var patternMatch = grants.patternMatch();
+        if (patternMatch != null) {
+            operations = patternMatch.lookup(action.resourceTypeClass(), action.resourceName());
```

Comment:
makes me wonder if there's some common interface here for looking up operations that could be implemented by TypePatternMatch and the TypeNameMap. So that we can stream over them rather than having this set of if/else/returns